File: 174-dungeon-game/174-dungeon-game.py
class Solution:
    def calculateMinimumHP(self, dungeon: List[List[int]]) -> int:
        row, col = len(dungeon), len(dungeon[0])
        hp = [1e9] * (col + 1)
        hp[col] = hp[col - 1] = 1
        print(hp)
        if row == 1 and col == 1 and dungeon[0][0] >= 0: return 1
        for i in range (row - 1, -1, -1):
            temp = [1e9] * (col + 1)
            for j in range (col - 1, -1, -1):
                minHP = -dungeon[i][j] + min(temp[j + 1], hp[j])
                temp[j] = 1 if minHP <= 0 else minHP
            hp = temp
        return hp[0]

File: 174-dungeon-game/174-dungeon-game.java
class Solution {
    public int calculateMinimumHP(int[][] dungeon) {
        int row = dungeon.length, col = dungeon[0].length;
        int[][] hp = new int[row + 1][col + 1];
        for(int[] row_elem : hp)
            Arrays.fill(row_elem, Integer.MAX_VALUE);
        hp[row][col - 1] = hp[row - 1][col] = 1;
        
        for(int i = row - 1; i >= 0; --i) {
            for(int j = col - 1; j >= 0; --j) {
                int minHP = -dungeon[i][j] + Math.min(hp[i][j + 1], hp[i + 1][j]);
                hp[i][j] = minHP <= 0 ? 1 : minHP;
            }
        }
        
        return hp[0][0];
    }
}